File: EventOrganizer.cs
// EventOrganizer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using QRCoder;

class Attendee
{
    [JsonPropertyName("id")] public string Id { get; set; }
    [JsonPropertyName("name")] public string Name { get; set; }
    [JsonPropertyName("email")] public string Email { get; set; }
    [JsonPropertyName("checked_in")] public bool CheckedIn { get; set; }
}

class Event
{
    [JsonPropertyName("id")] public string Id { get; set; }
    [JsonPropertyName("name")] public string Name { get; set; }
    [JsonPropertyName("date")] public string Date { get; set; }
    [JsonPropertyName("venue")] public string Venue { get; set; }
    [JsonPropertyName("attendees")] public List<Attendee> Attendees { get; set; } = new List<Attendee>();
}

class Organizer
{
    private List<Event> events = new List<Event>();
    private readonly string dataFile = "events.json";
    private readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    public Organizer() { Load(); }

    private void Load()
    {
        if (!File.Exists(dataFile)) return;
        string json = File.ReadAllText(dataFile);
        events = JsonSerializer.Deserialize<List<Event>>(json) ?? new List<Event>();
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(events, options);
        File.WriteAllText(dataFile, json);
    }

    private string GenerateId() => Guid.NewGuid().ToString().Substring(0,8);

    public void CreateEvent(string name, string date, string venue)
    {
        var e = new Event { Id = GenerateId(), Name = name, Date = date, Venue = venue };
        events.Add(e);
        Save();
        Console.WriteLine($"Event created: {e.Id} - {e.Name}");
    }

    public void ListEvents()
    {
        if (!events.Any()) { Console.WriteLine("No events."); return; }
        Console.WriteLine("\n📅 Events:");
        foreach (var e in events)
            Console.WriteLine($"  {e.Id}: {e.Name} ({e.Date}) at {e.Venue} - {e.Attendees.Count} attendees");
    }

    public void AddAttendee(string eventId, string name, string email)
    {
        var ev = events.FirstOrDefault(e => e.Id == eventId);
        if (ev == null) { Console.WriteLine($"Event {eventId} not found."); return; }
        var a = new Attendee { Id = GenerateId(), Name = name, Email = email, CheckedIn = false };
        ev.Attendees.Add(a);
        Save();
        Console.WriteLine($"Attendee {name} added with ticket ID {a.Id}");
    }

    public void ListAttendees(string eventId)
    {
        var ev = events.FirstOrDefault(e => e.Id == eventId);
        if (ev == null) { Console.WriteLine($"Event {eventId} not found."); return; }
        if (!ev.Attendees.Any()) { Console.WriteLine("No attendees."); return; }
        Console.WriteLine($"\n👤 Attendees for {ev.Name}:");
        foreach (var a in ev.Attendees)
        {
            string status = a.CheckedIn ? "✓" : "✗";
            Console.WriteLine($"  {a.Id}: {a.Name} ({a.Email}) - {status}");
        }
    }

    public void GenerateTickets(string eventId, string outputDir = ".")
    {
        var ev = events.FirstOrDefault(e => e.Id == eventId);
        if (ev == null) { Console.WriteLine($"Event {eventId} not found."); return; }
        Directory.CreateDirectory(outputDir);
        foreach (var a in ev.Attendees)
        {
            string data = $"{ev.Id}:{a.Id}";
            var qrGenerator = new QRCodeGenerator();
            var qrCodeData = qrGenerator.CreateQrCode(data, QRCodeGenerator.ECCLevel.M);
            var qrCode = new QRCode(qrCodeData);
            var qrImage = qrCode.GetGraphic(6);
            string filename = Path.Combine(outputDir, $"{a.Id}.png");
            qrImage.Save(filename, System.Drawing.Imaging.ImageFormat.Png);
            Console.WriteLine($"Ticket saved for {a.Name}: {filename}");
        }
    }

    public void Checkin(string eventId, string ticketId)
    {
        var ev = events.FirstOrDefault(e => e.Id == eventId);
        if (ev == null) { Console.WriteLine($"Event {eventId} not found."); return; }
        var a = ev.Attendees.FirstOrDefault(att => att.Id == ticketId);
        if (a == null) { Console.WriteLine($"Ticket {ticketId} not found for this event."); return; }
        if (a.CheckedIn) Console.WriteLine($"Attendee {a.Name} already checked in.");
        else { a.CheckedIn = true; Save(); Console.WriteLine($"Checked in {a.Name}."); }
    }

    public void Status(string eventId)
    {
        var ev = events.FirstOrDefault(e => e.Id == eventId);
        if (ev == null) { Console.WriteLine($"Event {eventId} not found."); return; }
        int total = ev.Attendees.Count;
        int checkedIn = ev.Attendees.Count(a => a.CheckedIn);
        Console.WriteLine($"\n📊 Event: {ev.Name}");
        Console.WriteLine($"  Total attendees: {total}");
        Console.WriteLine($"  Checked in: {checkedIn}");
        Console.WriteLine($"  Remaining: {total - checkedIn}");
        ListAttendees(eventId);
    }
}

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 1) { Console.WriteLine("Usage: EventOrganizer <command> [options]"); return; }
        var app = new Organizer();
        var cmd = args[0];
        switch (cmd)
        {
            case "event":
                if (args.Length < 2) { Console.WriteLine("event list | event create <name> <date> <venue>"); return; }
                if (args[1] == "list") app.ListEvents();
                else if (args[1] == "create" && args.Length == 5)
                    app.CreateEvent(args[2], args[3], args[4]);
                else Console.WriteLine("Invalid event subcommand.");
                break;
            case "attendee":
                if (args.Length < 3) { Console.WriteLine("attendee list <event_id> | attendee add <event_id> <name> <email>"); return; }
                if (args[1] == "list") app.ListAttendees(args[2]);
                else if (args[1] == "add" && args.Length == 5)
                    app.AddAttendee(args[2], args[3], args[4]);
                else Console.WriteLine("Invalid attendee subcommand.");
                break;
            case "ticket":
                if (args.Length < 3 || args[1] != "generate")
                { Console.WriteLine("ticket generate <event_id> [-o output_dir]"); return; }
                string outputDir = ".";
                for (int i=2; i<args.Length; i++)
                {
                    if (args[i] == "-o" && i+1 < args.Length)
                    { outputDir = args[i+1]; i++; }
                }
                app.GenerateTickets(args[2], outputDir);
                break;
            case "checkin":
                if (args.Length < 3) { Console.WriteLine("checkin <event_id> <ticket_id>"); return; }
                app.Checkin(args[1], args[2]);
                break;
            case "status":
                if (args.Length < 2) { Console.WriteLine("status <event_id>"); return; }
                app.Status(args[1]);
                break;
            default:
                Console.WriteLine("Unknown command.");
                break;
        }
    }
}

File: event_organizer.py
# event_organizer.py
import sys, os, json, uuid, argparse, datetime, qrcode
from PIL import Image, ImageDraw, ImageFont

DATA_FILE = "events.json"

class Event:
    def __init__(self, name, date, venue):
        self.id = str(uuid.uuid4())[:8]
        self.name = name
        self.date = date
        self.venue = venue
        self.attendees = []  # list of dicts with id, name, email, checked_in

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "date": self.date,
            "venue": self.venue,
            "attendees": self.attendees
        }

    @classmethod
    def from_dict(cls, data):
        e = cls(data["name"], data["date"], data["venue"])
        e.id = data["id"]
        e.attendees = data["attendees"]
        return e

class Organizer:
    def __init__(self):
        self.events = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.events = [Event.from_dict(e) for e in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([e.to_dict() for e in self.events], f, indent=2)

    def create_event(self, name, date, venue):
        ev = Event(name, date, venue)
        self.events.append(ev)
        self.save()
        print(f"Event created: {ev.id} - {name}")
        return ev

    def list_events(self):
        if not self.events:
            print("No events.")
            return
        print("\n📅 Events:")
        for e in self.events:
            print(f"  {e.id}: {e.name} ({e.date}) at {e.venue} - {len(e.attendees)} attendees")

    def add_attendee(self, event_id, name, email):
        ev = next((e for e in self.events if e.id == event_id), None)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        ticket_id = str(uuid.uuid4())[:8]
        ev.attendees.append({"id": ticket_id, "name": name, "email": email, "checked_in": False})
        self.save()
        print(f"Attendee {name} added to {ev.name} with ticket ID {ticket_id}")

    def list_attendees(self, event_id):
        ev = next((e for e in self.events if e.id == event_id), None)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        if not ev.attendees:
            print("No attendees.")
            return
        print(f"\n👤 Attendees for {ev.name}:")
        for a in ev.attendees:
            status = "✓" if a["checked_in"] else "✗"
            print(f"  {a['id']}: {a['name']} ({a['email']}) - {status}")

    def generate_tickets(self, event_id, output_dir="."):
        ev = next((e for e in self.events if e.id == event_id), None)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        os.makedirs(output_dir, exist_ok=True)
        for a in ev.attendees:
            # QR code data: event_id + ticket_id
            data = f"{ev.id}:{a['id']}"
            qr = qrcode.QRCode(box_size=10, border=2)
            qr.add_data(data)
            qr.make(fit=True)
            img = qr.make_image(fill="black", back_color="white").convert("RGB")
            # Add label with name and event
            # We'll create a new image with extra space for text
            ticket = Image.new("RGB", (img.width + 100, img.height + 60), "white")
            ticket.paste(img, (50, 10))
            draw = ImageDraw.Draw(ticket)
            try:
                font = ImageFont.truetype("/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf", 16)
            except:
                font = ImageFont.load_default()
            draw.text((50, img.height + 20), f"{a['name']} - {ev.name}", fill="black", font=font)
            filename = f"{output_dir}/{a['id']}.png"
            ticket.save(filename)
            print(f"Ticket saved for {a['name']}: {filename}")

    def checkin(self, event_id, ticket_id):
        ev = next((e for e in self.events if e.id == event_id), None)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        for a in ev.attendees:
            if a["id"] == ticket_id:
                if a["checked_in"]:
                    print(f"Attendee {a['name']} already checked in.")
                else:
                    a["checked_in"] = True
                    self.save()
                    print(f"Checked in {a['name']}.")
                return
        print(f"Ticket {ticket_id} not found for this event.")

    def status(self, event_id):
        ev = next((e for e in self.events if e.id == event_id), None)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        total = len(ev.attendees)
        checked = sum(1 for a in ev.attendees if a["checked_in"])
        print(f"\n📊 Event: {ev.name}")
        print(f"  Total attendees: {total}")
        print(f"  Checked in: {checked}")
        print(f"  Remaining: {total - checked}")
        self.list_attendees(event_id)

def main():
    parser = argparse.ArgumentParser(description="Event Organizer with QR Tickets")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    # Event create
    create_parser = subparsers.add_parser("event")
    create_sub = create_parser.add_subparsers(dest="subcmd", required=True)
    create_sub.add_parser("list")
    create_cmd = create_sub.add_parser("create")
    create_cmd.add_argument("name")
    create_cmd.add_argument("date")
    create_cmd.add_argument("venue")

    # Attendee
    attendee_parser = subparsers.add_parser("attendee")
    attendee_sub = attendee_parser.add_subparsers(dest="subcmd", required=True)
    attendee_sub.add_parser("list")
    add_cmd = attendee_sub.add_parser("add")
    add_cmd.add_argument("event_id")
    add_cmd.add_argument("name")
    add_cmd.add_argument("email")

    # Ticket
    ticket_parser = subparsers.add_parser("ticket")
    ticket_sub = ticket_parser.add_subparsers(dest="subcmd", required=True)
    gen_cmd = ticket_sub.add_parser("generate")
    gen_cmd.add_argument("event_id")
    gen_cmd.add_argument("-o", "--output", default=".")

    # Checkin
    checkin_parser = subparsers.add_parser("checkin")
    checkin_parser.add_argument("event_id")
    checkin_parser.add_argument("ticket_id")

    # Status
    status_parser = subparsers.add_parser("status")
    status_parser.add_argument("event_id")

    args = parser.parse_args()
    org = Organizer()

    if args.cmd == "event":
        if args.subcmd == "list":
            org.list_events()
        elif args.subcmd == "create":
            org.create_event(args.name, args.date, args.venue)
    elif args.cmd == "attendee":
        if args.subcmd == "list":
            org.list_attendees(args.event_id)
        elif args.subcmd == "add":
            org.add_attendee(args.event_id, args.name, args.email)
    elif args.cmd == "ticket":
        if args.subcmd == "generate":
            org.generate_tickets(args.event_id, args.output)
    elif args.cmd == "checkin":
        org.checkin(args.event_id, args.ticket_id)
    elif args.cmd == "status":
        org.status(args.event_id)

if __name__ == "__main__":
    main()
